### ISSUE TITLE
docker kill -s HUP <container> blows up a task

### ISSUE THREAD

**Issue by natefox** [aws/amazon-ecs-agent#335]
I'm running nginx:latest with consul-template. Each time consul-template updates the nginx config I need to HUP the container/nginx process so nginx will run the latest config (per the official docs under [Controlling NGINX](https://blog.docker.com/2015/04/tips-for-deploying-nginx-official-image-with-docker/)).

When I run `docker kill -s HUP <container id>` nginx reloads the config and all is well. However, ECS hates this. ECS-agent thinks the container has stopped - but it actually keeps running just fine. Then the task/container is reported as stopped and nginx is attempted to be relaunched (which it cant cause the existing container is holding the host port open).

Here's the ecs-agent log from when I run the `kill -s HUP` (with personal info removed):
```
2016-03-10T23:55:55Z [INFO] Adding event module="eventhandler" change="ContainerChange: arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 nginx -> STOPPED, Ports [{80 80 0.0.0.0 0}], Known Sent: RUNNING"
2016-03-10T23:55:55Z [INFO] Stopping container module="TaskEngine" task="mytask:5 arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22, Status: (RUNNING->STOPPED) Containers: [nginx (STOPPED->STOPPED),consul-template (RUNNING->STOPPED),]" container="consul-template(myregistry/consul-template) (RUNNING->STOPPED)"
2016-03-10T23:55:55Z [INFO] Sending container change module="eventhandler" event="ContainerChange: arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 nginx -> STOPPED, Ports [{80 80 0.0.0.0 0}], Known Sent: RUNNING" change="ContainerChange: arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 nginx -> STOPPED, Ports [{80 80 0.0.0.0 0}], Known Sent: RUNNING"
2016-03-10T23:55:55Z [INFO] Task change event module="TaskEngine" event="{TaskArn:arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 Status:STOPPED Reason: SentStatus:RUNNING}"
2016-03-10T23:55:55Z [INFO] Adding event module="eventhandler" change="ContainerChange: arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 consul-template -> STOPPED, Known Sent: RUNNING"
2016-03-10T23:55:55Z [INFO] Saving state! module="statemanager"
2016-03-10T23:55:55Z [INFO] Adding event module="eventhandler" change="TaskChange: arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 -> STOPPED, Known Sent: RUNNING"
2016-03-10T23:55:55Z [INFO] Sending container change module="eventhandler" event="ContainerChange: arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 consul-template -> STOPPED, Known Sent: RUNNING" change="ContainerChange: arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 consul-template -> STOPPED, Known Sent: RUNNING"
2016-03-10T23:55:55Z [INFO] Sending task change module="eventhandler" event="TaskChange: arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 -> STOPPED, Known Sent: RUNNING" change="TaskChange: arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22 -> STOPPED, Known Sent: RUNNING"
2016-03-10T23:55:55Z [INFO] Redundant container state change for task mytask:5 arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22, Status: (STOPPED->STOPPED) Containers: [nginx (STOPPED->STOPPED),consul-template (STOPPED->STOPPED),]: consul-template(myregistry/consul-template) (STOPPED->STOPPED) to STOPPED, but already STOPPED
2016-03-10T23:55:55Z [INFO] Redundant container state change for task mytask:5 arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22, Status: (STOPPED->STOPPED) Containers: [nginx (STOPPED->STOPPED),consul-template (STOPPED->STOPPED),]: consul-template(myregistry/consul-template) (STOPPED->STOPPED) to STOPPED, but already STOPPED
2016-03-10T23:55:55Z [INFO] Redundant container state change for task mytask:5 arn:aws:ecs:us-west-2:xxxxxxxxxxxx:task/8c119869-9ee3-467d-a467-c9a3c8193a22, Status: (STOPPED->STOPPED) Containers: [nginx (STOPPED->STOPPED),consul-template (STOPPED->STOPPED),]: consul-template(myregistry/consul-template) (STOPPED->STOPPED) to STOPPED, but already STOPPED
2016-03-10T23:56:05Z [INFO] Saving state! module="statemanager"
```

The biggest problem I have with this is if you have 2 containers in a task definition (nginx being one of them and marked essential), the second container is killed because it thinks nginx was killed. Due to this we're unable to run consul-template as a container along with nginx. We have to run 2 processes in one container and background nginx so it can be HUP'd properly.

**Comment by samuelkarp** [comment]
It looks like we're handling the `kill` event [incorrectly](https://github.com/aws/amazon-ecs-agent/blob/v1.8.1/agent/engine/docker_container_engine.go#L578).  In basic testing, `docker kill -s HUP` results in a `kill` event, `docker stop` results in `kill` followed by `die` and `stop`, while `docker kill` results in `kill` followed by `die`.  Before changing the logic here, we'll need to experiment with different Docker versions (and may need to just inspect the container to get accurate status).

**Comment by samuelkarp** [comment]
I've done more extensive testing across Docker versions and discovered a number of undocumented changes to the events.

| Docker version | `docker kill -s USR1` | `docker kill`  | `docker stop`                  | Container exit |
|----------------|-----------------------|----------------|--------------------------------|----------------|
| 1.5.0          | no event              | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.6.0          | no event              | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.6.1          | no event              | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.6.2          | no event              | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.7.0          | `kill`                | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.7.1          | `kill`                | `die`, `kill`  | `die`, `stop`                  | `die`          |
| 1.8.0          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.8.1          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.8.2          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.8.3          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.9.0          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.9.1          | `kill`                | `kill`, `die`  | `kill`, `kill`, `die`, `stop`  | `die`          |
| 1.10.0         | `kill`*               | `kill`, `die`* | `kill`, `kill`, `die`, `stop`* | `die`*         |
| 1.10.1         | `kill`*               | `kill`, `die`* | `kill`, `kill`, `die`, `stop`* | `die`*         |
| 1.10.2         | `kill`*               | `kill`, `die`* | `kill`, `kill`, `die`, `stop`* | `die`*         |
| 1.10.3         | `kill`*               | `kill`, `die`* | `kill`, `kill`, `die`, `stop`* | `die`*         |

\* These events are now typed with `container`.  Other events are emitted with type `network` and `volume`.

Given this table, it should be safe for us to ignore `kill`.  `die` looks to be consistently emitted when the container exits, and `stop` is emitted as well when `docker stop` is invoked.

**Comment by aaithal** [comment]
+1. That seems sane to me as well.

**Comment by samuelkarp** [comment]
Reported [upstream](https://github.com/docker/docker/issues/21331).

**Comment by samuelkarp** [comment]
This has been fixed in [1.8.2](https://github.com/aws/amazon-ecs-agent/releases/tag/v1.8.2), released today.
